### ISSUE TITLE
Stack overflow generating propnet for base.amazonsTorus_10x10

### ISSUE THREAD

**Issue by arr28** [SanchoGGP/ggp-base#386]
```
08:29:41.945 - Exception in thread "GamePlayer - Sancho 1.61d (9147)" 
08:29:41.945 - java.lang.StackOverflowError
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
08:29:41.965 -  at org.ggp.base.util.propnet.polymorphic.factory.OptimizingPolymorphicPropNetFactory.requiresDoes(OptimizingPolymorphicPropNetFactory.java:2047)
... (many times)
```

I thought I ran with an increased stack size but, having looked at it (`-d64 -Xms4g -Xmx5g -XX:+UseG1GC -XX:InitiatingHeapOccupancyPercent=80`), apparently not.   Perhaps just needs a `-Xss`.  Or perhaps a real infinite recursion bug.

**Comment by arr28** [comment]
Increasing the stack size not enough for this.  Even with excessively huge stacks (512MB, which could never be used for real because every thread gets that much stack pre-allocated), this still causes a stack overflow.

**Comment by arr28** [comment]
The propnet for this game has loops as evidenced by a log reached if the `requiresDoes` processing is disabled.

```
19:59:02.942 ? Propnet loops detected - unable to check for duplicate logic amongst components fed from such loops
```

The propnet loops cause `requiresDoes` to recurse infinitely, eventually leading to stack overflow.  `requiresDoes` is only used for propnet pruning.  It's always functionally safe for it to return `false` (according to existing comments in that routine).  So, possible solutions include...
- Impose arbitrary depth limit on `requiresDoes`.  This is crude, but I've tried it and it works.  Risk is that I set a bad limit (I tried 1000) and end up not doing pruning when we need it.
- Keep track of the components already processed.  If we see the same one again, bail out.
- Bring forward whatever processing it is the produces the log above.  Don't even try `requiresDoes` if we know there are loops.
